src/transport/SecureSessionTable.cpp

```diff
@@ -0,0 +1,274 @@
+            }
+            else
+            {
+                (void) fabricIndex;
+            }
+        }
+    }
+
+    SecureSession * result = mEntries.CreateObject(*this, secureSessionType, localSessionId, localNodeId, peerNodeId, peerCATs,
+                                                   peerSessionId, fabricIndex, config);
+    return result != nullptr ? MakeOptional<SessionHandle>(*result) : Optional<SessionHandle>::Missing();
+}
+
+Optional<SessionHandle> SecureSessionTable::CreateNewSecureSession(SecureSession::Type secureSessionType,
+                                                                   ScopedNodeId sessionEvictionHint)
+{
+    Optional<SessionHandle> rv = Optional<SessionHandle>::Missing();
+    SecureSession * allocated  = nullptr;
+
+    auto sessionId = FindUnusedSessionId();
+    VerifyOrExit(sessionId.HasValue(), rv = Optional<SessionHandle>::Missing());
+
+    //
+    // We allocate a new session out of the pool if we have space in it. If we don't, we need
+    // to run the eviction algorithm to get a free slot. We shall ALWAYS be guaranteed to evict
+    // an existing session in the table in normal operating circumstances.
+    //
+    if (mEntries.Allocated() < CHIP_CONFIG_SECURE_SESSION_POOL_SIZE)
+    {
+        allocated = mEntries.CreateObject(*this, secureSessionType, sessionId.Value());
+        VerifyOrExit(allocated != nullptr, rv = Optional<SessionHandle>::Missing());
+    }
+    else
+    {
+        allocated = EvictAndAllocate(sessionId.Value(), secureSessionType, sessionEvictionHint);
+    }
+
+    rv             = MakeOptional<SessionHandle>(*allocated);
+    mNextSessionId = sessionId.Value() == kMaxSessionID ? static_cast<uint16_t>(kUnsecuredSessionId + 1)
+                                                        : static_cast<uint16_t>(sessionId.Value() + 1);
+
+exit:
+    return rv;
+}
+
+SecureSession * SecureSessionTable::EvictAndAllocate(uint16_t localSessionId, SecureSession::Type secureSessionType,
+                                                     const ScopedNodeId & sessionEvictionHint)
+{
+    VerifyOrDieWithMsg(!mRunningEvictionLogic, SecureChannel,
+                       "EvictAndAllocate isn't re-entrant, yet someone called us while we're already running");
+    mRunningEvictionLogic = true;
+
+    ChipLogProgress(SecureChannel, "Evicting a slot for session with LSID: %d, type: %u", localSessionId,
+                    (uint8_t) secureSessionType);
+
+    VerifyOrDie(mEntries.Allocated() <= CHIP_CONFIG_SECURE_SESSION_POOL_SIZE);
+
+    //
+    // Create a temporary list of objects each of which points to a session in the existing
+    // session table, but are swappable. This allows them to then be used with a sorting algorithm
+    // without affecting the sessions in the table itself.
+    //
+    Platform::ScopedMemoryBufferWithSize<SortableSession> sortableSessions;
+    sortableSessions.Calloc(CHIP_CONFIG_SECURE_SESSION_POOL_SIZE);
```

Comment:
Does `ForEachSession` effectively give us a null check for `sortableSessions`?  If not, suggest:

```suggestion
    sortableSessions.Calloc(CHIP_CONFIG_SECURE_SESSION_POOL_SIZE);
    if (!sortableSessions)
    {
        return nullptr;
    }
```

Reply:
Actually, the other thing I need to do here is to only Calloc as many sessions as we actually have currently allocated.